### PR TITLE
Add native user constraints to control matchers more finely

### PR DESCRIPTION
**Pull request by nicolasvasilache**

### CODE REVIEW

File: lib/Dialects/LinalgTransform/Transforms/PDL.cpp [hunk 1]
```diff
@@ -59,6 +63,138 @@ static void noOpRewriter(ArrayRef<PDLValue> args, ArrayAttr constantParams,
...
+}
+
+/// Check whether the unique Operation* `op` stored in `value` (assumed) is
+/// equivalent to the unique StringRefAttr operation name passed in
+/// `constantParams`.
+/// Equivalence is achieved when either:
+///   1. `op` has the name stored in `constantParams`.
+///   2. `op` and `constantParams` are both linalg ops and their structured
+///      interfaces as well as their bodies are equivalent.
+///      Structured interfaces equivalence is a simple ttribute level check.
```
Comment (ftynse) [inline lib/Dialects/LinalgTransform/Transforms/PDL.cpp:141]: Typo: attribute

File: lib/Dialects/LinalgTransform/Transforms/PDL.cpp [hunk 2]
```diff
@@ -59,6 +63,138 @@ static void noOpRewriter(ArrayRef<PDLValue> args, ArrayAttr constantParams,
...
+//      `divisor` IntegerAttr entries.
+// Succeed if `value`[`operand_number`] is a ranked type whose `dim` is a
+// multiple of `divisor`.
+static LogicalResult isDimMultipleOf(PDLValue value, ArrayAttr constantParams,
+                                     PatternRewriter &rewriter) {
+  auto operands = value.cast<ValueRange>();
+  auto dict = constantParams.begin()->cast<DictionaryAttr>();
+  auto dim = dict.getAs<IntegerAttr>("dim").getInt();
+  auto divisor = dict.getAs<IntegerAttr>("divisor").getInt();
+  auto operandNumber = dict.getAs<IntegerAttr>("operand_number").getInt();
```
Comment (ftynse) [inline lib/Dialects/LinalgTransform/Transforms/PDL.cpp:188]: All these look like they they can fail the typecheck and crash. Could we return failure + have debug output on why?

File: python/examples/matmul/test_schedule_dialect.py [hunk 3]
```diff
@@ -24,22 +25,30 @@
...
     sequence = transform.SequenceOp()
     with ir.InsertionPoint(sequence.body.blocks[0]):
-      transform.TileOp(pdl_pattern_name, sizes=[12, 32, 1], pad=False)
+      matched = transform.MatchOp(pdl_pattern_name)
+      transform.TileOp(matched, sizes=[12, 32, 1], pad=False)
       transform.BufferizeOp()
       transform.LowerVectorsOp()
       transform.LowerToLLVMOp()
 
+  print(module)
```
Comment (ftynse) [inline python/examples/matmul/test_schedule_dialect.py:50]: Leftover debug output

**Review by ftynse — APPROVED** [review]

File: lib/Dialects/LinalgTransform/Transforms/PDL.cpp [hunk 4]
```diff
@@ -59,6 +63,240 @@ static void noOpRewriter(ArrayRef<PDLValue> args, ArrayAttr constantParams,
...
+///        b. a more advanced version would "subtract" the bodies and fold, cse
+///           and canonicalize to fixed point. If the result is "all zeros",
+///           then the bodies would be equivalent (really isomorphic).
+///   3. other cases TBD (e.g. vector.generic when available).
+static LogicalResult isEquivalentToOp(PDLValue value, ArrayAttr constantParams,
+                                      PatternRewriter &rewriter) {
+  auto maybeOp = value.dyn_cast<Operation *>();
+  if (!maybeOp)
+    return failure(); // TODO: notifyMatchFailure needs an Operation* handle.
+  Operation *op = *maybeOp;
```
Comment (giuseros) [inline lib/Dialects/LinalgTransform/Transforms/PDL.cpp:154]: I am trying this right now :) This is supposed to be `Operation *op = maybeOp;`, right?

**Comment by giuseros** [comment]
Given the correction in the comment, I tried it out and it is working like  a charm :) I am trying out a similar approach for element-wise transformations. Is that ok?

File: lib/Dialects/LinalgTransform/Transforms/PDL.cpp (hunk 4 above)
Comment (nicolasvasilache) [reply lib/Dialects/LinalgTransform/Transforms/PDL.cpp:154]: Fantastic, the more the merrier.
Next, for my part, I'll connect this to the tuner.

**Pull request merged**